extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -0,0 +1,661 @@
+                            Math.floor(diff / (1000 * 60 * 60 * 24));
+                        const ncve = 'NoncurrentVersionExpiration';
+                        if (applicableRules[ncve].NoncurrentDays &&
+                        applicableRules[ncve].NoncurrentDays <= daysSinceCR) {
+                            this.sendObjectEntry({
+                                action: 'delete',
+                                target: {
+                                    owner: bucketData.owner,
+                                    bucket: bucketData.bucket,
+                                    key: version.Key,
+                                    version: data.VersionId,
+                                },
+                            });
+                        }
+                        return next();
+                    });
+                },
+            ], err => cb(err));
+        }, done);
+    }
+
+    // _compareDeleteMarkers(bucket, rules, deleteMarkers, versions, log, cb) {
+    //      // will be completed later
+    // }
+
+    _compareWithLCRules(bucketData, bucketLCRules, data, log, done) {
+        async.parallel([
+            next => this._compareContents(bucketData, bucketLCRules,
+                data.Contents, log, next),
+            next => this._compareVersions(bucketData, bucketLCRules,
+                data.Versions, log, next),
+            next => this._compareDeleteMarkers(bucketData, bucketLCRules,
+                data.DeleteMarkers, data.Versions, log, next),
+        ], done);
+    }
+
+    /**
+     * Process a single bucket entry with lifecycle configurations enabled
+     * @param {array} bucketLCRules - array of bucket lifecycle rules
+     * @param {object} bucketData - bucket data from zookeeper bucketTasksTopic
+     * @param {string} bucketData.bucket - name of bucket
+     * @param {string} bucketData.ownerID - owner id
+     * @param {string} [bucketData.prefix] - prefix
+     * @param {string} [bucketData.nextKeyMarker] - next key marker for
+     *   versioned buckets
+     * @param {string} [bucketData.nextVersionIdMarker] - next version id marker
+     *   for versioned buckets
+     * @param {string} [bucketData.nextContinuationToken] - next continuation
+     *   token marker for non-versioned buckets
+     * @param {function} done - callback(error, data)
+     * @return {undefined}
+     */
+    processBucketEntry(bucketLCRules, bucketData, done) {
+        const log = this.log.newRequestLogger();
+
+        log.debug('processing bucket entry', {
+            bucket: bucketData.bucket,
+            owner: bucketData.ownerId,
+        });
+
+        // NOTE:
+        //  With the addition of mpu's, there was no easy way to handle
+        //  how to process entries w/ markers. Will need to rethink this and
+        //  refactor
+
+        // get all objects whether versioned or non-versioned
+        async.parallel([
+            cb => {
+                if (bucketData.versionIdMarker ||
```

Comment:
Are `versionIdMarker` and `uploadIdMarker` mutually exclusive? If so, you can get rid of `async.parallel` and pass them the `done` callback directly.

Reply:
Yes, they are mutually exclusive, but there is the case when we receive a new bucket task that has none of these markers defined, in which case, I want to go ahead and process both mpu's and (non-version && version). When paginating through markers, a bucket entry will have only one marker defined, in which case, we process for that one marker type and ignore others